docs/config.rst
.. _config:

Configuration
=============

There are a number of options to configure how Wazimap behaves. Options are
set in ``settings.py`` in the ``WAZIMAP`` dict.

``name``
  The full name of the website. Default: ``Wazimap Example``

``url``
  The site's base url, including scheme, WITHOUT and ending slash.
  If your site supports HTTPS, this **must** be an ``https`` url, otherwise
  embeds that are over HTTPS might not load correctly.
  Default: ``http://wazimap.example``

``twitter``
  Twitter handle. Default: ``@Wazimap``

``ga_tracking_id``
  Google Analytics tracking id.

``country_code``
  Two-letter country code for the country this site relates to.

``google_geocode_api_key``
  Google Geocoding API key, if you want to use more than the free limit.

``cache_secs``
  How many seconds should cacheable Wazimap pages be cached for? Default: ``60 * 60``

``embed_cache_secs``:
  How many seconds should Wazimap embed pages be cached for? Default: ``24 * 60 * 60``

``geodata``
  The dotted-path of the class to use for geo data helper routines.
  See :ref:`geos` for more info.
  Default: ``wazimap.geo.GeoData``

``levels``
  Geography levels. This must be a dict similar to the following: ::

      {
        'country': {
          'name': 'country',      # optional, defaults to the level code
          'plural': 'countries',  # plural version of the name
          'children': ['province'],
        },
        ...
      }

  Default: ::

       'country': {
           'plural': 'countries',
           'children': [],
       }

``comparative_levels``
  Ordered list of levels to compare a place with; should generally be only 2-3 levels.
  Default: ``[]``

``geometry_data``
  Map from geometry names to local static files with geometry data. This is used
  to both draw maps in webpages and for geolocation.
  Wazimap uses Django's static files finder to find these files, so the
  files must be stored in your application's `static` directory.

  For maps, topojson is preferred because it's smaller. But for geolocation,
  because it's done server side, it requires geojson (shapely doesn't understand
  topojson). Ideally you should have both `.geojson` and `.topojson` files
  and specify the `.topojson` version here. Wazimap will look for geojson equivalents
  of topojson files.

  The top level of this dict MUST be the geo version. If you're not using versioned
  geographies, use the empty string.

  If a level isn't specified here, Wazimap will try to find a file under the key ``''``.

  Default: ::

      {
        '': {
          '': 'geo/all.geojson',
        }
      }

``map_centre``, ``map_zoom``
  Centre coordinates and zoom level defaults for maps. Centre must be a ``[lat, long]`` pair
  and zoom is a zoom level (1-12).
  If not set, both are determined from the geometry.

``embed_url``
  The full url, including scheme (but no ending slash) where embeds are hosted.
  This is generally not used.  If your site supports HTTPS, this **must** be an
  ``https`` url, otherwise embeds that are over HTTPS might not load correctly.
  The default is the same as ``url``.

``na_label``
  The label to display for stats which have no data available.
  The default is N/A.

``default_geo_version``
  Default geo version to use when loading geographies. If ``None``,
  the most recent version in the Geography's table (``geo_data.latest_version``) is used.

``legacy_embed_geo_version``
  The geo version to use for legacy embeds that don't specify a geo version.
  If ``None``, uses the latest version (``geo_data.latest_version``).
  If you're introducing versioned geographies and your users have already embedded charts,
  you probably want to set this to your earliest version so that embeds continue showing the original data.

``primary_release_year``
  The primary release year to use for each geo level. The default is to use
  the ``latest`` release. Set this if you have newer releases at some geo
  levels, such as a 2010 national census down to the city level, and a 2015
  partial census to the provincial level.

Localisation
------------

You can customise how Wazimap formats numbers shown on a profile page by
using Django's `FORMAT_MODULE_PATH <https://docs.djangoproject.com/en/1.10/ref/settings/#std:setting-FORMAT_MODULE_PATH>`_ setting.
These settings in particular apply:

``THOUSAND_SEPARATOR``
  The thousand separator used when formatter numbers.

``DECIMAL_SEPARATOR``
  The decimal separator used when formatter numbers.

``CURRENCY_SYMBOL``
  The currency symbol to be used before currency amounts. Not in Django.
  Default: $

Additional Options
------------------

These options are specific to Wazimap and are set at the top-level of your settings.py file, not in the ``WAZIMAP`` dict.

``STRIP_WWW``
  Set this to ``True`` to redirect users visiting your site at ``www.example.com`` to ``example.com``. This is the reverse of
  Django's `PREPEND_WWW <https://docs.djangoproject.com/en/1.10/ref/settings/#prepend-www>`_ setting.
  Default is ``False``.


CHANGES.rst
Wazimap Version History
=======================

2.0.0 (?)

This release of Wazimap introduces some backwards incompatible changes. You WILL need to follow the upgrade steps.
Be sure to upgrade to 1.x before upgrading to 2.x.

* Wazimap now supports multiple releases of a dataset. This means that you can, for example, have two releases of a national census in Wazimap at the same time. You will need to make some changes to how you build profile pages.
* BREAKING: SimpleTable and FieldTable are now Django models.
* BREAKING: Table management is now done through the Django admin interface. You must upgrade your existing tables.
* BREAKING: ``get_objects_by_geo`` has been replaced by ``FieldTable.get_rows_for_geo``.

1.1.1 (21 November 2017)
-----------------------

* FIX another bug for versioned geos for static maps


1.1.0 (23 October 2017)
-----------------------

* FIX support for versioned geos for static maps

1.0.1 (21 September 2017)
-------------------------

This release of Wazimap introduces some backwards incompatible changes.

* Support versioned geographies.
* BREAKING: The Geography model has changed. You must run ``python manage.py migrate`` when updating.
* BREAKING: The method signature of the ``profile_builder`` function has changed. It is now called as ``get_profile(geo, profile_name, request)``. You will need to update your profile methods to work with a ``geo`` object rather than a ``level`` and a ``code``. Use ``geo.geo_level`` and ``geo.geo_code`` instead.
* BREAKING: Some methods on the ``GeoData`` object now take a ``Geography`` object rather than a ``geo_level`` and a ``geo_code``.
* BREAKING: The ``geography_data`` setting must include a geography version at the top of the dict. Use ``''`` by default. eg: ``{'': {'country': 'geo/country.geojson'}}``
* Some methods on the ``GeoData`` object now take an optional ``version`` parameter.
* A new method ``GeoData.get_comparative_geos`` makes it easier to build the profile page for comparative geographies.
* FIX: load streetmap tiles over https if necessary
* FEATURE: site menu
* FEATURE: use SCSS stylesheets, support customising using SCSS variables.
* Attribution updates to reflect OpenUp, the new name for Code for South Africa.

0.8.2 (20 April 2017)
---------------------

* Make the 'show data / embed' links for charts more visible

0.8.1 (18 April 2017)
---------------------

* Fix metadata for field tables that re-use db tables

0.8.0 (12 April 2017)
---------------------

* Support FieldTables that share a common database table
* Remove support for ``FieldTable.table_per_level`` parameter.

0.7.6 (3 April 2017)
--------------------

* Tell robots not to crawl /api/

0.7.5 (15 March 2017)
---------------------

* Remove unnecessary Auth, Admin, Messages and Sessions middleware. If you need these middleware, add them to your INSTALLED_APPS and MIDDLEWARE_CLASSES in your settings file.
* New setting ``STRIP_WWW`` to redirect ``www.example.com`` to ``example.com``, turned off by default.

0.7.4 (13 March 2017)
---------------------

* Allow overriding of header block in map view

0.7.3 (9 March 2017)
--------------------

* Add a stub help page at /help

0.7.2 (3 March 2017)
--------------------

* Don't exclude partial matches from google geocoding results

0.7.1 (25 January 2017)
-----------------------

**Important** This version makes GDAL an optional install. You must ensure you add GDAL to your requirements file, or install wazimap using ``wazimap[gdal]]``. See http://wazimap.readthedocs.io/en/latest/deploying.html

* Make GDAL and Shapely optional. This makes it much easier to get up and running with Wazimap without getting GDAL installed, which can be tricky.
* Support GDAL 1.x or 2.x (2.x not fully tested).

0.6.1 (11 January 2017)
-----------------------

* FIX bug in testing database names

0.6.0 (11 January 2017)
-----------------------

* Support null values in a table. This allows you to indicate to Wazimap that you don't have data for a particular field/indicator in a particular place. It is still necessary to ensure that all places have an entry in the database for all indicators.
* Automated build-time testing on `Travis <https://travis-ci.org/Code4SA/wazimap>`_.

0.5.8 (21 November 2016)
------------------------

* Support percentage values in the map view

0.5.7 (17 November 2016)
------------------------

* Support tables with raw values that are percentages
* Version 0.5.6 was not released

0.5.5 (7 November 2016)
-----------------------

* Fix embedded charts (contributed by mevey)

0.5.4 (28 October 2016)
-----------------------

* Don't show error for empty FieldTable with has_total=True

0.5.3 (21 October 2016)
-----------------------

* Fix loadGeometryForGeoIds bug (contributed by cliftonmcintosh)

0.5.2 (28 September 2016)
-------------------------

* Fix profile page tables for non-percentages

0.5.1 (27 September 2016)
-------------------------

* Correctly send comparative_levels setting to profile page

0.5.0 (22 September 2016)
-------------------------

* Support floating point values in tables

0.4.6 (20 September 2016)
-------------------------

* Be more tolerant of GDAL versions >= 1.11.0 and < 2.0
* Lazy load OGR/GDAL
* Don't allow Django 1.10

0.4.4 (8 September 2016)
------------------------

* Allow overriding of favicon.

0.4.2 (23 August 2016)
----------------------

* Support decimal places in chart tables

0.4.1 (22 August 2016)
----------------------

* FEATURE control the number of decimals for stats shown with the ``_stat_list`` include by using the ``decimals`` parameter.
* FEATURE adjust the currency symbol used for ``currency`` stats shown by ``_stat_list`` by setting ``CURRENCY_SYMBOL`` in the file specified by Django's FORMAT_MODULE_PATH.
* FEATURE get_stat_data supports percent_grouping and slices

0.3.3 (17 August 2016)
----------------------


wazimap/settings.py
# Django settings for Wazimap
import os
import sys
import dj_database_url


DEBUG = os.environ.get('DJANGO_DEBUG', 'true') == 'true'
TEMPLATE_DEBUG = DEBUG
TESTING = 'test' in sys.argv[1:3]

ROOT_URLCONF = 'wazimap.urls'
WSGI_APPLICATION = 'wazimap.wsgi.application'

ADMINS = (('Greg Kempe', 'greg@code4sa.org'),)
MANAGERS = ADMINS

DATABASE_URL = os.environ.get('DATABASE_URL', 'postgresql://wazimap:wazimap@localhost/wazimap')
DATABASES = {
    'default': dj_database_url.parse(DATABASE_URL),
}
DATABASES['default']['ATOMIC_REQUESTS'] = True

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.humanize',
    'django.contrib.staticfiles',
    'sass_processor',
    'wazimap.apps.WazimapConfig',
    'census',
]

ALLOWED_HOSTS = ['*']
TIME_ZONE = 'Africa/Johannesburg'
LANGUAGE_CODE = 'en-za'
USE_I18N = False
USE_L10N = True
USE_THOUSAND_SEPARATOR = True
USE_TZ = True

# SECURITY WARNING: keep the secret key used in production secret!
if DEBUG:
    SECRET_KEY = '!%j-u4&(q8qu4@dq=ukth27+q!v-!h^jck14bf=spqht847$4q'
else:
    SECRET_KEY = os.environ.get('DJANGO_SECRET_KEY')


API_URL = None


# Static files
MEDIA_ROOT = ''
MEDIA_URL = ''

STATIC_ROOT = 'static'
STATIC_URL = '/static/'

STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    'sass_processor.finders.CssFinder',
)


# Templates
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
)
TEMPLATE_CONTEXT_PROCESSORS = (
    'django.core.context_processors.media',
    'django.core.context_processors.request',
    'django.core.context_processors.static',
    'django.contrib.auth.context_processors.auth',
    'census.context_processors.api_url',
    'wazimap.context_processors.wazimap_settings',
)


MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'corsheaders.middleware.CorsMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'wazimap.middleware.RedirectMiddleware',
    # 'django.middleware.clickjacking.XFrameOptionsMiddleware',
)

# CORS
CORS_ORIGIN_ALLOW_ALL = True
CORS_ALLOW_METHODS = (
    'GET',
    'OPTIONS'
)


# Databases and caches
if DEBUG:
    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.dummy.DummyCache',
        }
    }
else:
    CACHES = {
        'default': {
            'BACKEND': 'django.core.cache.backends.filebased.FileBasedCache',
            'LOCATION': '/var/tmp/wazimap_cache',
        }
    }


LOGGING = {
    'version': 1,
    'disable_existing_loggers': True,
    'formatters': {
        'verbose': {
            'format': '%(asctime)s %(levelname)s %(module)s %(process)d %(thread)d %(message)s'
        },
    },
    'handlers': {
        'console': {
            'level': 'DEBUG',
            'class': 'logging.StreamHandler',
            'formatter': 'verbose'
        },
    },
    'loggers': {
        '': {
            'handlers': ['console'],
            'level': 'ERROR',
        },
        'census': {
            'level': 'DEBUG' if DEBUG else 'INFO',
        },
        'django': {
            'level': 'DEBUG' if DEBUG else 'INFO',
        },
        'django.template': {
            'level': 'ERROR',
        },
        'wazimap': {
            'level': 'DEBUG' if DEBUG else 'INFO',
        },
    }
}


# Wazimap-specific settings

# Redirect www.foo.com to foo.com? This is the reverse of Django's PREPEND_WWW
STRIP_WWW = False

WAZIMAP = {
    # The full name of the website
    'name': 'Wazimap',

    # The site's base URL, including scheme.
    # NOTE: if your site supports HTTPS, this MUST be
    # an 'https' url, otherwise embeds that are over HTTPS might
    # not load correctly.
    'url': 'http://wazimap.example',

    # Twitter handle
    'twitter': '@Wazimap',

    # Google Analytics tracking id (ua-xxxxx-x)
    'ga_tracking_id': '',

    # Two-letter country code for the country this site relates to
    'country_code': '',

    # Google Geocoding API key, if you want to use more than the free limit
    'google_geocode_api_key': '',

    # How many seconds should cacheable Wazimap pages be cached for?
    'cache_secs': 60 * 60,

    # How many seconds should Wazimap embed pages be cached for?
    'embed_cache_secs': 24 * 60 * 60,

    # the dotted-path of the class to use for geo data helper routines
    'geodata': 'wazimap.geo.GeoData',

    # Geography levels. This must be a dict similar to the following:
    #
    # {
    #   'country': {
    #     'name': 'country',      # optional, defaults to the level code
    #     'plural': 'countries',  # plural version of the name
    #     'children': ['province'],
    #   },
    #   ...
    # }
    'levels': {
        'country': {
            'plural': 'countries',
            'children': [],
        }
    },

    # ordered list of levels to compare a place to; should generally be 2-3 levels only.
    'comparative_levels': [],

    # Map from geometry names to local static files with geometry data. This is used
    # to both draw maps in webpages and for geolocation.
    # Wazimap uses Django's static files finder to find these files, so the
    # files must be stored in your application's `static` directory.
    #
    # For maps, topojson is preferred because it's smaller. But for geolocation,
    # because it's done server side, it requires geojson (shapely doesn't understand
    # topojson). Ideally you should have both `.geojson` and `.topojson` files
    # and specify the `.topojson` version here. Wazimap will look for geojson equivalents
    # of topojson files.
    #
    # If a level isn't specified here, Wazimap will try to find a file under the key `''`.
    'geometry_data': {
        '': {
            '': 'geo/all.geojson',
        },
    },

    # centre coordinates and zoom level defaults for maps. Centre must be a ``[lat, long]`` pair
    # and zoom is a zoom level (1-12).
    # If not set, the centre is determined from the geometry.
    'map_centre': None,
    'map_zoom': None,

    # URL where embeds are hosted, default is the site URL
    # NOTE: if your site supports HTTPS, this MUST be
    # an 'https' url, otherwise embeds that are over HTTPS might
    # not load correctly.
    'embed_url': None,

    # Label to display if data is not available.
    'na_label': 'N/A',

    # Default geo version to use when loading geographies. If None,
    # uses the most recent version of the Geography. This is useful if your
    # geography table contains more recent geographies, but you don't want
    # to use the latest when the request doesn't specifify a version.
    'default_geo_version': None,

    # The geo version to use for legacy embeds that don't specify a geo version.
    # If None, uses the latest version.
    # If your users have already used embeds and you're introducing versioned
    # geographies, you probably want to set this to your earliest version, so
    # that embeds continue to show the original data.
    'legacy_embed_geo_version': None,

    # The primary release year to use for each geo level. The default is to use
    # the `latest` release. Set this if you have newer releases at some geo
    # levels, such as a 2010 national census down to the city level, and a 2015
    # partial census to the provincial level.
    'primary_release_year': {},
}
